### PR TITLE
Refactor sprite renderer

### PR DESCRIPTION
**Pull request by singlecoder**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
**Refactor sprite renderer.**

### What is the current behavior? (You can also link to an open issue here)
https://github.com/oasis-engine/engine/issues/32
### What is the new behavior (if this is a feature change)?

- [ ] Add some new api.
- [ ] Optimize the underlying rendering pipeline.

### CODE REVIEW

File: packages/core/src/2d/sprite/Sprite.ts [hunk 1]
```diff
@@ -0,0 +1,260 @@
+import { Vector2, Vector4 } from "@oasis-engine/math";
+import { RefObject } from "../../asset/RefObject";
+import { Engine } from "../../Engine";
+import { Texture2D } from "../../texture";
+
+// A 2D Rectangle defined by X and Y position, width and height.
+export class Rect {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:7]: I think this should put in math folder and named Rect or Rectange

File: packages/core/src/2d/sprite/Sprite.ts [hunk 2]
```diff
@@ -0,0 +1,260 @@
+import { Vector2, Vector4 } from "@oasis-engine/math";
+import { RefObject } from "../../asset/RefObject";
+import { Engine } from "../../Engine";
+import { Texture2D } from "../../texture";
+
+// A 2D Rectangle defined by X and Y position, width and height.
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:6]: Write comments according to specifications

File: packages/core/src/2d/sprite/Sprite.ts [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+   * @remarks
+   * x=left, y=right, z=top, w=bottom.
+   */
+  private _border: Vector4 = new Vector4();
+  /** Location of the sprite's center point in the rect on the original texture, specified in pixels. */
+  private _pivot: Vector2 = new Vector2();
+  /** Location of the sprite on the original texture, specified in pixels. */
+  private _rect: Rect = new Rect();
+  /** The array containing sprite mesh triangles. */
+  public triangles: number[] = [];
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:46]: public property should ahead of private

File: packages/core/src/2d/sprite/Sprite.ts [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   * The reference to the used texture.
+   */
+  get texture(): Texture2D {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:164]: get/set should ahead of constructor

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+import { Entity } from "../../Entity";
+import { Material } from "../../material";
+import { Renderer } from "../../Renderer";
+import { Sprite } from "./Sprite";
+
+/**
+ * Draw mode.
+ * @readonly
+ */
+export enum DrawMode {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:12]: enum should write in independent file

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]

File: packages/core/src/RenderPipeline/SpriteMaterial.ts [hunk 6]
```diff
@@ -1,40 +1,40 @@
...
 
-import { Shader } from "@oasis-engine/core";
+import { Shader } from "../shader";
 
 const spriteVertShader = `
 precision highp float;
 
-uniform mat4 matProjection;
-uniform mat4 matView;
+uniform mat4 u_projMat;
```
Comment (GuoLei1990) [inline packages/core/src/RenderPipeline/SpriteMaterial.ts:8]: you can use "viewProject" uniform to improve performance

File: packages/core/src/RenderPipeline/SpriteBatcher.ts [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+import { Texture2D } from "../texture";
+import { Shader } from "../shader";
+import { ShaderData } from "../shader/ShaderData";
+import { Renderer } from "../Renderer";
+
+export class SpriteBatcher {
+  /** The maximum number of vertices. */
+  private static MAX_VERTICES: number = 256;
+
+  private _batchedQueue;
```
Comment (GuoLei1990) [inline packages/core/src/RenderPipeline/SpriteBatcher.ts:24]: Write Type

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]

File: packages/core/src/2d/sprite/Sprite.ts (hunk 1 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/Sprite.ts]: > I think this should put in math folder and named Rect or Rectange

done

File: packages/core/src/2d/sprite/Sprite.ts (hunk 2 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/Sprite.ts]: > Write comments according to specifications

done，has removed rect from sprite to math

File: packages/core/src/2d/sprite/Sprite.ts (hunk 3 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/Sprite.ts]: > public property should ahead of private

done

File: packages/core/src/2d/sprite/Sprite.ts (hunk 4 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/Sprite.ts]: > get/set should ahead of constructor

done

File: packages/core/src/2d/sprite/SpriteRenderer.ts (hunk 5 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/SpriteRenderer.ts]: > enum should write in independent file

done

File: packages/core/src/RenderPipeline/SpriteBatcher.ts (hunk 7 above)
Comment (singlecoder) [reply packages/core/src/RenderPipeline/SpriteBatcher.ts]: > Write Type

done

File: packages/core/src/RenderPipeline/SpriteMaterial.ts (hunk 6 above)
Comment (singlecoder) [reply packages/core/src/RenderPipeline/SpriteMaterial.ts]: > you can use "viewProject" uniform to improve performance

done

File: packages/core/src/2d/enums/DrawMode.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Draw mode.
+ * @readonly
```
Comment (GuoLei1990) [inline packages/core/src/2d/enums/DrawMode.ts:3]: Why write readonly at enum.

File: packages/core/src/2d/sprite/Sprite.ts [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+  private static _VER_UV_TRI_FLAG = 0x7;
+
+  /** The array containing sprite mesh triangles. */
+  public triangles: number[] = [];
+  /** The base texture coordinates of the sprite mesh. */
+  public uv: Vector2[] = [new Vector2(), new Vector2(), new Vector2(), new Vector2()];
+  /** The array containing sprite mesh vertex positions. */
+  public vertices: Vector2[] = [new Vector2(), new Vector2(), new Vector2(), new Vector2()];
+
+  private _tempVec2: Vector2 = new Vector2();
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:20]: temp vector  use static is better

File: packages/core/src/2d/sprite/Sprite.ts [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    }
+  }
+
+  /**
+   * TODO
+   * The border sizes of the sprite.
+   * @remarks
+   * x=left, y=right, z=top, w=bottom.
+   */
+  get border(): Vector4 {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:76]: If not implement, you can delete and add later.

File: packages/core/src/2d/sprite/Sprite.ts [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+    if (!pivot) {
+      pivot = new Vector2(0.5, 0.5);
+    }
+
+    if (rect.x + rect.width > texture.width || rect.y + rect.height > texture.height) {
+      throw new Error("rect out of range");
+    }
+
+    // Init.
+    this._dirtyFlag = Sprite._VER_UV_TRI_FLAG;
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:158]: const value init in construtor is not good

File: packages/core/src/2d/sprite/Sprite.ts [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+      triangles[4] = 0;
+      triangles[5] = 3;
+    }
+  }
+
+  /**
+   * Update mesh data of the sprite.
+   * @returns True if the data is refreshed, false otherwise.
+   */
+  updateMeshData(): boolean {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:229]: This is internal function or public function,internal should use _XX and add internal comment

File: packages/core/src/2d/sprite/SpriteMaterial.ts [hunk 13]
```diff
@@ -1,42 +1,41 @@
...
 {
   // Only use the Alpha of the texture as a mask, so that the tint color can still be controlled to fade out.
-  vec4 baseColor = texture2D(s_diffuse, v_uv);
+  vec4 baseColor = texture2D(u_texture, v_uv);
   gl_FragColor = baseColor * v_color;
 }
 `;
 
-Shader.create("Sprite", spriteVertShader, spriteFragmentShader);
+Shader.create("Sprite-Default", spriteVertShader, spriteFragmentShader);
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteMaterial.ts:41]: The shader name "Sprite" is better than "Sprite-Default"

File: packages/core/src/2d/enums/DrawMode.ts (hunk 8 above)
Comment (singlecoder) [reply packages/core/src/2d/enums/DrawMode.ts]: > Why write readonly at enum.

has deleted

File: packages/core/src/2d/sprite/Sprite.ts (hunk 9 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/Sprite.ts]: > temp vector use static is better

done

File: packages/core/src/2d/sprite/Sprite.ts (hunk 10 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/Sprite.ts]: > If not implement, you can delete and add later.

has deleted

File: packages/core/src/2d/sprite/Sprite.ts (hunk 11 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/Sprite.ts]: > const value init in construtor is not good

done

File: packages/core/src/2d/sprite/Sprite.ts (hunk 12 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/Sprite.ts]: > This is internal function or public function,internal should use _XX and add internal comment

done

File: packages/core/src/2d/sprite/SpriteMaterial.ts (hunk 13 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/SpriteMaterial.ts]: > The shader name "Sprite" is better than "Sprite-Default"

done

File: packages/core/src/2d/sprite/Sprite.ts [hunk 14]
```diff
@@ -0,0 +1,232 @@
+import { Vector2, Vector4, Rect } from "@oasis-engine/math";
+import { RefObject } from "../../asset/RefObject";
+import { Engine } from "../../Engine";
+import { Texture2D } from "../../texture";
+
+export class Sprite extends RefObject {
+  private static _tempVec2: Vector2 = new Vector2();
+
+  private static _VERTICES_FLAG = 0x1;
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:9]: You can  write a internal enum in this  file.

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+    const needUpdate = sprite._updateMeshData();
+    const { triangles, uv, vertices, texture } = sprite;
+
+    // Update vertices position in world space.
+    const posZ = transform.position.z;
+    for (let i = 0; i < 4; ++i) {
+      const curVertex = vertices[i];
+      const tempPos = SpriteRenderer._tempVec4;
+      tempPos.setValue(curVertex.x, curVertex.y, posZ, 1);
+      Vector4.transform(tempPos, modelMatrix, tempPos);
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:79]: use "Vector3.transformCoordinate"

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+
+  /**
+   * The current draw mode of the Sprite Renderer.
+   */
+  get drawMode(): DrawMode {
+    return this._drawMode;
+  }
+
+  // TODO
+  set drawMode(value: DrawMode) {}
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:127]: You can delete, and add later

File: packages/core/src/2d/sprite/Sprite.ts [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+  private static _TRIANGLES_FLAG = 0x4;
+  /** Sprite._VERTICES_FLAG | Sprite._UV_FLAG | Sprite._TRIANGLES_FLAG */
+  private static _VER_UV_TRI_FLAG = 0x7;
+
+  /** The array containing sprite mesh triangles. */
+  public triangles: number[] = [];
+  /** The base texture coordinates of the sprite mesh. */
+  public uv: Vector2[] = [new Vector2(), new Vector2(), new Vector2(), new Vector2()];
+  /** The array containing sprite mesh vertex positions. */
+  public vertices: Vector2[] = [new Vector2(), new Vector2(), new Vector2(), new Vector2()];
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:20]: For triangles、vertices、uv is not safe API，developer will change the value and broken your cache

File: packages/core/src/2d/sprite/Sprite.ts [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+  }
+
+  /**
+   * Location of the sprite's center point in the rect on the original texture, specified in pixels.
+   */
+  get pivot(): Vector2 {
+    return this._pivot;
+  }
+
+  set pivot(value: Vector2) {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:71]: If develop change the pivot and then call vertices still get old data until next farme.

File: packages/core/src/2d/sprite/Sprite.ts (hunk 17 above)
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:20]: "Vertex" include postion and other vertex element，if this data is just position, how about call positions

File: packages/core/src/2d/sprite/Sprite.ts (hunk 14 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/Sprite.ts]: > You can write a internal enum in this file.

done

File: packages/core/src/2d/sprite/SpriteRenderer.ts (hunk 16 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/SpriteRenderer.ts]: > You can delete, and add later

done

File: packages/core/src/2d/sprite/SpriteRenderer.ts (hunk 15 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/SpriteRenderer.ts:75]: > use "Vector3.transformCoordinate"

Vector4. transformCoordinate

File: packages/core/src/2d/sprite/Sprite.ts [hunk 19]
```diff
@@ -0,0 +1,233 @@
+import { Rect, Vector2 } from "@oasis-engine/math";
+import { RefObject } from "../../asset/RefObject";
+import { Engine } from "../../Engine";
+import { Texture2D } from "../../texture";
+
+export class Sprite extends RefObject {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:6]: Write comments

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+    const { _triangles: triangles, _uv: uv, _positions: positions, texture } = sprite;
+
+    // Update vertices position in world space.
+    const posZ = transform.position.z;
+    for (let i = 0; i < 4; ++i) {
+      const curVertex = positions[i];
+      const tempPos = SpriteRenderer._tempVec3;
+      tempPos.setValue(curVertex.x, curVertex.y, posZ);
+      Vector3.transformToVec3(tempPos, modelMatrix, tempPos);
+      _positions[i].setValue(tempPos.x, tempPos.y, tempPos.z);
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:76]: You can write ```Vector3.transformToVec3(tempPos, modelMatrix, _positions[i]);``` to improve performance

File: packages/core/src/2d/sprite/Sprite.ts (hunk 19 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/Sprite.ts:9]: > Write comments

done

File: packages/core/src/2d/sprite/SpriteRenderer.ts (hunk 20 above)
Comment (singlecoder) [reply packages/core/src/2d/sprite/SpriteRenderer.ts]: > You can write `Vector3.transformToVec3(tempPos, modelMatrix, _positions[i]);` to improve performance

done

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+
+        if (flipY) {
+          curPos.y = y * 2 - curPos.y;
+        }
+      }
+
+      this._setDirtyFlagFalse(SpriteRenderer._FLIP_FLAG);
+    }
+
+    // @ts-ignore
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:168]: Can delete

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+        );
+        Vector3.transformToVec3(localVertexPos, worldMatrix, _positions[i]);
+      }
+
+      this._setDirtyFlagFalse(SpriteRenderer._FLIP_FLAG);
+      this._setDirtyFlagFalse(SpriteRenderer._SPRITE_FLAG);
+      this._isWorldMatrixDirty.flag = false;
+    } else if (this._isContainDirtyFlag(SpriteRenderer._FLIP_FLAG)) {
+      const flipX = this._isContainDirtyFlag(SpriteRenderer._FLIP_X_FLAG);
+      const flipY = this._isContainDirtyFlag(SpriteRenderer._FLIP_Y_FLAG);
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:150]: Here has a BUG, If developer call ```flipY = true```, then call ```flipY = false```. Im fact you do not need to re-compute,but.........

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 23]
```diff
@@ -0,0 +1,197 @@
...
+  private static _tempVec3: Vector3 = new Vector3();
+  private static _defaultMaterial: Material = null;
+
+  private static _FLIP_X_FLAG = 0x1;
+  private static _FLIP_Y_FLAG = 0x2;
+  private static _SPRITE_FLAG = 0x4;
+  /** SpriteRenderer._FLIP_X_FLAG | SpriteRenderer._FLIP_Y_FLAG */
+  private static _FLIP_FLAG = 0x3;
+  /** SpriteRenderer._FLIP_X_FLAG | SpriteRenderer._FLIP_Y_FLAG | SpriteRenderer._SPRITE_FLAG */
+  private static _ALL_FLAG = 0x7;
```
[the remark this answers is not in the document]
Comment (singlecoder) [reply packages/core/src/2d/sprite/SpriteRenderer.ts]: > You can use internal enum

done

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+    const { sprite } = this;
+    if (!sprite) {
+      return;
+    }
+
+    const { _positions } = this;
+    const { transform } = this.entity;
+
+    // Update sprite data.
+    const needUpdate = sprite._updateMeshData();
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:110]: how `needUpdate` call `localDirty` or `localUpdate`

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+      this._cacheFlipY = flipY;
+    } else if (this._isContainDirtyFlag(DirtyFlag.Flip)) {
+      const { flipX, flipY } = this;
+      const flipXChange = this._cacheFlipX !== flipX;
+      const flipYChange = this._cacheFlipY !== flipY;
+
+      if (flipXChange || flipYChange) {
+        const { x, y } = transform.worldPosition;
+
+        for (let i = 0, len = _positions.length; i < len; ++i) {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:142]: `len` use `n` is better and `++i` use `i++`, why not use 4 directly.

File: packages/core/src/2d/sprite/SpriteRenderer.ts [hunk 26]
```diff
@@ -0,0 +1,207 @@
...
+  constructor(entity: Entity) {
+    super(entity);
+    this._isWorldMatrixDirty = entity.transform.registerWorldChangeFlag();
+  }
+
+  /**
+   * Push the render data of the sprite to render queue.
+   * @param camera - Camera which is rendering
+   */
+  _render(camera: Camera): void {
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/SpriteRenderer.ts:100]: add internal and do not need other comments

File: packages/core/src/2d/sprite/Sprite.ts [hunk 27]
```diff
@@ -0,0 +1,236 @@
...
+
+    if (rect.x + rect.width > texture.width || rect.y + rect.height > texture.height) {
+      throw new Error("rect out of range");
+    }
+
+    // Init.
+    this.texture = texture;
+    this.rect = rect;
+    this.textureRect = rect;
+    this.pivot.setValue(pivot.x * rect.width, pivot.y * rect.height);
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:138]: Can't understand this design.

File: packages/core/src/2d/sprite/Sprite.ts [hunk 28]
```diff
@@ -0,0 +1,236 @@
...
+    engine: Engine,
+    texture: Texture2D,
+    rect: Rect = null,
+    pivot: Vector2 = null,
+    pixelsPerUnit: number = 100
+  ) {
+    super(engine);
+
+    if (!rect) {
+      rect = new Rect(0, 0, texture.width, texture.height);
```
Comment (GuoLei1990) [inline packages/core/src/2d/sprite/Sprite.ts:124]: use setValue and cloneTo

**Review by GuoLei1990 — APPROVED** [review]
+1

**Pull request merged**